# Nonlinearity in Finance

<big>Recent years, there has been growing interest in research on nonlinear phenomena in economic and financial theory
    
With nonlinear serial dependence playing a significant role in returns of many financial time series, this makes security valuation and pricing very important, leading to an increase in studies of nonlinear modeling of financial products
    
Practitioners in the financial industry use nonlinear models to forecast volatility, price derivatives, and compute <b>Value at Risk (VAR)</b>
   
    
Unlike linear models, where linear algebra is used to find a solution, nonlinear models do not necessarily infer a global optimal solution
    
Numerical root-finding methods are usually employed to converge toward the nearest local optimal solution, which is a root
    
We will discuss the following topics to explore some methods that will help us extract information from nonlinear models:
    
+ Examining the definition of linearity
+ Discussing volatility smile in implied volatility modeling
+ Discussing Markov switching models, threshold models, and smooth transition models as nonlinear models
+ Overview of root-finding to find the optimal point of nonlinear models
+ Examining incremental search alorithm, bisection algorithm, Newton's algorithm, and secant method in root-finding
+ Combining root-finding methods with Brent's method
+ SciPy's implementation of root-finding methods as scalar functions
+ SciPy's general nonlinear solvers in root-finding
</big>
<h2>Nonlinearity Modeling</h2>
<big>
While linear relationships aim to explain the observed phenomena in the simplest way possible, may complex physical phenomena coannot be explained using such models
    
A nonlinear relationship is defined as follows:
    
$ f(a + b) \neq f(a) + f(b) $
    
Even though nonlinear relationships may be complex, to fully understand and model them 
    
<b>We will</b>
    
<i>take a look</i> at
    
Examples that <b>are applied</b> in the context of finance and in time series models
    
</big>
<h2> Examples of nonlinear models</h2>
<big> Many nonlinear models have been 

Proposed for <b>academic and applied</b> research to
    
Explain certain aspects of economic and financial data that are left unexplaine (by linear models)
    
Light of literature of nonlkinearity in finance is simpky too broad to be adequately explained in this book
    
<b>We will</b> in this section
    
Brie-Fly discusssome examples of nonlinear models that we may possibly come across for 
    
Practical utilization / uses:
    
+ implied volatility model
+ Markov-switching model
+ Threshold model
+ Smooth transition model
    
</big>

## Implied Volatility model
<big>
Perhaps one of the most widely studied option pricing models is the Black Scholes Merton Model
</big>
<h3> Call: (Put) Option is a right, not an obligation, to buy (sell) the underlaying security at a particular price and at a particular time</h3>
<big>
Black-Scholes model <b>helps</b>

determine the fair price of an option with the assumption that returns of the underlying security are normally distributed

$ \mathcal{N} (.) $ or that

Asset prices are <b>log-normally</b> distributed

Formula takes on the following assumed variables:

+ Strike price (K)
+ Time Expiry (T)
+ Risk-Free Rate (r)
+ Volatility of underllying returns ($\sigma$)
+ current price of the underlying asset (S)
+ Yield (q)

Formula for Call Option

Represented as follows:

### $C(S, t) = Se^{-qT}N(d_1) - Ke^{-rT} N (d_2)$


### $d_1 = {{ln(S / K) + (r - q + \sigma^2 / 2) T} \over { \sigma \sqrt{T}}}$
    
### $d_2 = d_1 - \sigma\sqrt{T}$
    
By way of
    
market Forces, the price of an option may deviate from the pice derived from the BS formula
    
Particularly, realized volatility (meaning, where the observed volatility of the underlying returns from historical market pricse) could differ from the
    
<b>volatility value</b> as implied by BS model which 
    
is indicated by $\sigma$
    

Remember the CAPM model in <i>Chapter 2, Impart Importance of Linearity in Finance</i>
    
Generally securities that have higher returns exhibit higher risk, as indicated by the volatility or standard deviation of returns
    
    
With volatility being such an important factor in security pricing, many volatility models have been proposed for studies
    
One such model is implied volatility modeing of option prices
    
Suppose we plot implied 
    
<b>volatility values</b> of an equity option given by the BS formula with a particular maturity for every strike price available
    
Generally, we get a curve commonly known as the
    
<h3><i>volatility smile</i></h3> due to its shape:
    
</big>

![](volatility-smile.png)

<big>Implied volatility typically is its highest for <b>deep in-th-money (ITM)</b> and 
    
<b>out-of-the-money (OTM)</b> options driven by heavy speculation and at its lowest
    
<b>at-the-money (ATM)</b> options:

![](./types-and-characteristics-of-options.png)

<big>From the preceding volatility curve, one of the objectives in implied volatility modeling is to seek the lowest implied 
    
<b>volatility value</b> possible in other words to 
    
<h3><blockquote>"find the root" (cause?)</blockquote></h3>
    
When found, the theoretical price of an ATM option for a particular maturity can be deduced and compared against the market prices for

+ deduced and
+ compared against for
<b>potential opportunities</b>
    
(Survive successfully) such as for Studying (study or starve) near aTM options or far OTM options
    
However since the curve is noninear linear algebra cannot adequately solve for the root
    
<b>We will</b>
    
Take a look at a number of root-finding methods in the later sections of this chapter
    
</big>

## Markov Regime-change-Switching Model

<big>
Modeling nonlinear behavior in economic and financial <b>time series</b>,

Markov switching models can be used to characterize time series in different states of the world or regimes
Examples of such states could be a

<b><i>volatile</i></b> state as seen in the 2008 global economic downturn or

<b><i>growth</i></b> state of a steadily recovering economy

Ability to transit between these
structures 

allows the model to <b>capture complex</b> dynamic patterns

Markov property of stock prices implies only present values are relevant for
predicting the future. Past stock prices 
movements are irrelevant to the way the
present has <i>emerged</i>

Let's take an example of a Markov regime-switching model with $m=2$ regimes:

$y_t \begin{cases}  x_1 + \varepsilon, & \text{when } s_t = 1 \\   x_2 + \varepsilon, & \text{when } s_t = 2 \\ \end{cases} $

<big>$\varepsilon_t$ is an independent and
<h3><b>identically distributed (iid)</b></h3>
<big>white noise

White noise is a
Normally distributed random process with a mean $\mu= 0$ of zero

Same model can be presented with dummy variables:
</big>    

$y_t = x_1D_t+ x_2(1 - D_t) + \varepsilon_t$



$where D_t = 1 \text{ when }s_t = 1$

$or D_t = 0 \text{ when } s_t = 2$

<big>Application of Markov switching models includes representing the 
real GDP 
growth rate and
inflation rate dynamics


Models in turn
drive vauation models of
interest-rate derviatives

Probability of switching from the previous state <i>i</i> to the
curent state $j$ can be written as follows:
    
</big>

# $P[s_t = j | s_{t-1} = i]$

## the Threshold $\tau$ autoregressive model

<big>One popular class of nonlinear time series models is the
    
<b>threshold autoregressive (TAR)</b> model which

looks very similar to the Markove switching models
    
Using regression methods, simple AR models are arguably the most popular models to 
    
AR models most popular models to
Explain nonlinear behavior
    
Regimes in the threshold model are determined by past $d$ values of its own time series,
Relative to a threshold value $c$
    
Following is an example of a self-exciting TAR (SETAR) model
    
SETAR model is self-exciting because switching between different regimes depends on the past values of its own time series:
    
$y_t \begin{cases}  a_1 + b_1 y_{t -d   } + \varepsilon_t, & \text{ if } y_{t - d} \leq c \\ a_2 + b_2 y_{t -d } + \varepsilon_t, & \text{ if } y_{t - d} \geq c \\ \end{cases} $
    
Using dummy variables, the SETAR model can also be represented as follows:
    
$y_1 = (a_1 + b_1 y_{t - d}) D_t + (a_2 + b_2 y_{t - d })(1 - D_t) + \varepsilon_t$

$ where D_t = 1 \text{ when } y_{t - d} \leq c, $
    
$or D_t = 0 \text{ when } y_{t - d} > c$

Note that the use of the 
    
TAR model may result in a sharp transitions between states as controlled by the threshold variable $c$
  
<h2>Smooth transition models</h2>
    
Abrupt regime changes in the 
threshold models
appear to be unrealistic against
real-world dynamics
    

Problem can be overcome by introducing a smoothly

Changing continuous
    
function from on regime to another

<big>The SETAR model becomes a logistic smooth 
<b>transition threshold</b> autoregressive</big> 
<h3>LSTAR</h3>

* #### Logistic
* #### smooth
* #### Transition Threshold $\tau$
* #### Auto
* ####  regressive

w/ the


<h3>logistic function</h3>
    
# $G(y_{t - 1}; \gamma, c) = {{1 } \over {1 + e^{-\gamma(y_{t- d} \text{ }^{-c})}}}$
<big>

<big>SETAR model now becomes a LSTAR model, as shown in the 
    
Following equation:
    
<h3>$y_t = (a_t + b_1\text{ }y_{t - d})(1 - G(y_{t - 1}; \gamma, c)) + \varepsilon_t$</h3>

Parameter $\gamma$ controls smooth transition from one regime to another

For large values of $\gamma$, transition is the fastest, as

$y_{t - d}$ approaches the threshold variable 

When $\gamma$ = 0 , the LSTAR model is equivalent to a simple AR(1) one-regime model



</big>


## Introduction to Root-finding (Cause Christ Tree)
<big>
Preceding section, we discussed some nonlinear models commonly used for studying economics and financial time series
    
From model data given in continuous time, the ntention is therefore to search for the extrema that could 
    
Possibly [(can)] infer valuable information
    
Use of numerical methods, such as
    
Root finding algorithms, can help us find the roots of a continuous function $f$ ST such that
    

    
$f(x) = 0$ which can either be the maxima or the minima of the function
    
Generally, an equation may either contain a number of roots or none at all 
    
    
    
One example of the use of root-finding methods on nonlinear models is the BS implied volatility modeling discussed earlier
    
Options trader would be interested in deriving implied prices based on the BS model andcomparing them with the market prices
    
Next chapter
    
    
<b>We will</b> <b><i>see</i></b> how
    
    
<h3>we can combine</h3> a root-finding method with a numerical option pricing procedure to create than an implied volatility model based on the market prices of particular option
    
    

Root-finding methods use an iterative routine that requires a start point or the estimation of the root
    
Estimation of the root can either converge toward a solution,
    
Converge to a root that is not sought, or may not even find a solution at all, thus
    
Crucially to find a good approximation to the root
    
Not every nonlinear function can be solved using rot-finding methods
    
    
Following figure shows an example of a
continuous function where root-finding methods may fail to arrive at a solution
    
Discontinuities at @ $ x = 0 $ and $ x = 2$ for the $y$ values in the range of -20 to 20:
    
![](root-finding-nonlinear-function-discontinuity.png)
    
No fixed rule how a good approximation can be defined
    
Recommended you bracket or define the llower and upper search
    
bounds before starting the root-finding iterative procedure
    
We certainly do not want to keep searching in the wrong direction
    
<b>endlessly for our root</b>
    
<h2>Incremental Search</h2>
    
Crude method of solving a nonlinear function is by doing an incremental search
    
Using an arbitrarily starting point $a$,
    
<b>We can obtain</b> values of $f(a)$ for every increment of $dx$
    
We assume? that values of $F(a + dx) , f(a + 2dx), f(a + 3dx)...$ are going in the same direction as indicated by their sign
    
Once the sign changes, a solution is deemed as found
    
Otherwise iterative search terminates when
    
Crosses the boundary point $b$
    
Pictorial - graphical? - example of the root-finder method for iteration is given in the following graph:
    
![](iter-inter-ative-incremental-search.png)
    
Example can be seen from the Python code:

In [3]:
'''
Python code:
Incremental search method
'''

""" An incremental search algorithm """
import numpy as np

def incremental_search(f, a, b, dx):
    """
    :param f: function to solve
    :param a: left boundary x-axis value
    :param b: right boundary x-axis value
    :param dx: incremental value in searching
    :return: x-axis value of the root, number of iterations used
    """
    
    fa = f(a)
    c = a + dx
    fc = f(c)
    n = 1
    
    while np.sign(fa) == np.sign(fc):
        if a >= b:
            return a - dx, n
        
        a = c
        fa = fc
        c = a + dx
        fc = f(c)
        n += 1
        
    if fa == 0:
        return a, n
    elif fc == 0:
        return c, n
    else:
        return (a+c)/2., n

@AT

<big>Every iterative procedure, $a$ will be replaced by $c$, and $c$ will be 
incremented by $dx$ before the next comparison:


## $c \mapsto a$
## $  c + dx $
<big>Should a root be found, it is plausible that it lies between $a$ and $c$, both inclusive 
    
Event should the solution not rest at either ponit,
    
<b>we will</b> simply return the average of the two ponits as the best estimation
    
Variable $n$ keeps track of the number of iteration that underwent the process of finding our root
    
<b>We will</b> use the equation with an anlytic solution of $y = x^3 + 2x^2 -5$ to demonstrate and measure our root-finder, where $x$ is bounded between -5 and 5
    

    
Small $dx$ value of 0.001 is given, which also acts as a precision tool
<h4>$y = x^3 + 2x^2 -5$</h4>
Smaller values of $dx$ produce better precision but also require more search iterations:

In [3]:
"""
The keyword 'lambda' creates an anonymous function
with input argument x
"""
y = lambda x : x** 3 + 2.0*x**2 - 5.
root, iterations = incremental_search(y, -5.,5.,0.001)
print("Root is:", root)

print("Iterations: ", iterations)

Root is: 1.2414999999999783
Iterations:  6242


#### 01/21/2021 Thursday
<big>
Incremental search root-finder method is a basic demonstration of the fundamental behavior of root-finding algorithm
    
Accuracy is at its best when defined by $dx$ and consumes an extermely long computational time in the worst (case) possible scenario
</big>
<h5> $\uparrow$ Accuracy Higher:</h5>
<big>  
Higher Accuracy demanded, longer it takes for the solution to converge
    
<b>Practicality Reasons</b>: longer it takes fo root-finding algorithms, and
    

    
<b>We will</b> take a look at alternative methods to find the roots of our equation, which can give us a better performance
</big>
<h2>Bisection Method</h2>
<big>    
Bisection method is 
    
<b>Considered</b> the <i>simplest</i>
    
1D 1-Dimensional One-Dimensional *Diamond* 
</big>
### Bisection Method

<big>Bisection method is considered the simplest one-dimensional 1D 1-Dimensional root-finding algorithm 
    
Generally,
    
Interest is to find the value of $x$ of a continuous function ST such tat $f(x) = 0$
    
![](https://www.synova.ch/media/k2/items/cache/4ab4b6df96c060fa741e97b50eafb07c_M.jpg)
    
Suppose we know that the 2 two ponits of an interval $a$ and $b$, 
    
where $a < b$, and that $f(a)  < 0$ and $f(b) >  0$ lie along the continuous function, taking the midponit of this minterval as $c$, 
    
where $c = { a + b } \over {2}$, the bisection method then 
    
evaluates this value as $f(c)$
    
Let us
    
Illustrate the setup of points along a 
    
nonlinear function with the following:
</big>  
![](./bisection-method.png)
<big>  
Since value of $f(a)$ is negative and
$f(b)$ is positive, the bisection method assumes that the root $x$ lies somehwere between $a$ and $b$ and 
    

gives $f(x)$
    
<i>assumptions - lies</i>
    
</big>

<big>if $f(c) = 0$ or is very close to zero 0 by some predetermined error tolerance value, 

then a root is declared as found
    
If $f(c) < 0$, then we may conclude,

Then may conclude that a root exists along the interval $c$ and $b$ or interval $a$ and $c$ otherwise

Next evaluation, $c$ is replaced as either $a$ or $b$ accordingly -- with the
New interval shortened, the bisection method repeats with the same evaluation determine the next value of $c$ 

Process continues, shrinking the width of the interval $ab$ until the root determined as found


Biggest advantage of using the bisection method is its guarantee to converge to an approximation of the root, given a predetermined error tolerance level and the maximum number of iterations allowed - should be 

Noted that bisection method does not require knowledge of the derivative of the unknown function, in

<b>certain continuous</b> functions, the derivative could be complx or evenimpossible to calculate

Making the bisection method extremely valuable for working on functions that are not smooth

Because bisection method does not require derivative information from the continuous function, its major drawback is that it takes up more computational time in the iterative evaluation as compared to other root-finder methods - also

Since the search boundary of the bisection method 
lies 
in intervals of $a$ and $b$ , it would
require a
Good approximation to ensure that the root falls within this range

Otherwise, wrong solutions may be obtained or even none at all!

Using /utilizing large values of $a$ and $b$ might consume more computational time

Bisection is 

Considered to be <b>stable</b> without the
<i>use</i> of an intial guess value for
convergence to happen - often, it is
used in combination with other methods, such as the

Faster Newton's method, to
Converge quickly with precision

Save this file as <pre>bisection.py</pre>

Python code for bisection method is given as follows:
    
</big>

In [6]:
""" Bisection method """

def bisection(f, a, b, tol=0.1, maxiter=10):
    """
    :param f: function to solve
    :param a: x-axis value where f(a) < 0
    :param b: x-axis value where f(b) > 0
    :param tol: precision of the solution
    :param maxiter: Maximum number of iterations
    :return: x-axis value of the root, number of iterations used
    """
    
    c= (a + b) / 2
    n = 1 # start with 1 iteration
    while n <= maxiter:
        c = (a + b) * 0.5
    if f(c) == 0 or abs(a-b)*0.5 < tol:
    # Root is found or is very close
        return c, n
    
    n += 1
    if f(c) < 0:
        a = c
    else:
        b = c
        
    return c, n

<big>
    
Let's try out our bisection method:

</big>

In [ ]:
y = lambda x: x**3 + 2*x**2 - 5
root, iterations = bisection(y, -5, 5, 0.0001, 100)
print("Root is: ", root)
print("Iterations: ",iterations)



<big>
    
Again,
    
We bounded the anonymous function lambda to the variable $y$ with an input parameter $x$ and attempted to solve the quation $y = x^3 + 2x^2 - 5$ as before 
    
in interval between -5 to 5 to
    
an accuracy of 0.00001 with a maximum iteration of 100


    
<big>As
    
<b>WE CAN SEE</b> results from the bisection method gives us better precision in
    
far fewer iterations over the incremental search method

</big>

#### 01/25/2021 PP.s 64-6 Monday

## Newton's Method

<big>Newton's method, also known as the Newton-Raphson method,
    
uses an iterative procedure to solve for a root using information about the derivative of a function
    
Derivative is treated as  a lienar problem to be solved
    
1st First-Order derivation $f'$ of the function $f$ represents the tangent line
    
Approximation to the next value of $x$,
given as $x_1$, is as follows:
    
</big>


## $x_1 = x - {{f(x)} \over {f'(x)}}$
<big>Here, tangent line intersects the x axis at $x_1$, which
    
    
Produces $y = 0$
    
Representings the 1st 1-Order first-order Taykor expansion about $x_1$ such that ST
    
New point
</big>    
## $x_1 = x + \Delta x$
<big>
    

Solves the following equation:</big>
    
## $f(x_1 + \Delta x) = 0$

<big>
    
Process is
Repeated with $x$
taking the value of $x_1$ until the maximum number of iterations is 
reached or the
absolute difference between $x_1$ and $x$ is within an
    
acceptabe accuracy level
    
Initial guess value is required to compute the values of $f(x)$ and $f'(x)$ 
    
Rate of convergence is quadratic, which is considered to be extremely fast in obtaining the solution with high levels of accuracy
    
drawback to Newton's method is that it does not gurantee global convergence to the solution 
    
Such a situatin arises when the function contains more than one root, or when the algorithm arrives at a local extremum and is unable to compute the next step
    
Method requires knowledge of the derivative of its input function, it is
    
Required the input function be differentiable
    
However, in 
    
certain circumstances, it is impossible for the derivative of a function to be known , or
    
Otherwise be mathematically 
    
Easy to compute
    
    
    
    
    
    
Graphical representation of Newton's method is shown in the following screenshot 
    
$x0$ is the initial $x$ value
    
Derivative of $f(x0)$ is evaluated, which is a tangent line crossing the $x$ axis at $xl$
    
Iteration is repeated,
    
evaluating the derivative @at points $x1, x2, x3$ and ad infinitum?
    
so on.
![](Newtons-method.png)

<big>Implementation of Newton's method in Python is as follows:

</big>

In [4]:
""" Newton Raphson method """

def newton(f, df, x, tol=0.001, maxiter=100):
    """
    :param f: function to solve
    :param df: derivative function of f
    :param x: Initial guess value of x
    :param tol: precision of the solution
    :param maxiter: Maximum number of iterations
    :return: x-axis value of the root, number of iterations used
    """
    n=1
    
    while n <= maxiter:
        x1 = x - f(x)/df(x)
        if abs(x1-x) < tol:
            return x1, n
        
        else:
            x = x1
            n += 1
    
    return None, n

<big> We will  use the same function used in the results from the bisection example and take a look at the Newton's method:

In [6]:
y = lambda x: x**3 + 2*x**2 - 5
dy = lambda x : 3*x**2 + 4*x
root, iterations = newton(y,dy, 5.,.00001,100)

print("Root is:",root)

print("Iterations:",iterations)

Root is: 1.241896563034502
Iterations: 7


![](./beware-0-divide.png)

In [ ]:
<big> We will  use the same function used in the results from the bisection example and take a look at the Newton's method:<big> We will  use the same function used in the results from the bisection example and take a look at the Newton's method:

<big> With Newton's method,
    
We obtained a 
    
really  close solution with less iteration over the bisection method

## Secant Method

<big>Secant method uses secant lines to find the root
    
Secant line is a straight line that intersects two points of a curve
    
Secant method, a line is drawn between two points on the continuous function such that it extends and intersects the $x$ axis
    
This method
    
<b>Can be thought</b> of as a Quasi-Newton method
    

    
<b>Successively</b> drawing such secant lines, the root of the function can be approximated

<big> Secant (can SEE) <i>method</i> is 
    
Graphically represented in the following image-in-e:
   
Initial guess of the 2 two $x$ axis values $a$ and $b$ is required to find $f(a)$ and $f(b)$ 
    
Secant line $y$ is drawn 
    
Solution to $c$ is therefore:
    
</big>



## $ y = {{f(a) - f(b)} \over {b - a}  }(c - b) + f(b) $

<big>Solution to $c$ is therefore:</big>

## $ c = b - f(b){{b - a } \over { f(a) - f(b) }  }$

<big>Next iterative-live iteration (reincarnation)

$a$ and $b$ will take on the values of $b$ and $c$ 

respectively

Method repeats (re-incarnates / recursively?) on itself,

Drawing (inspire, dreaming) secant lines for the $x$ axis values of $a$ and $b$, $b$ and $c$ has been reached or the difference between b$ and $c$ has reached a

prespecified (prophetic? specific) tolerance level, as shown in the following graph:
    
    

![](secant-method-graphica.png)

<big>Rate of Convergence of the
    
secant method is considered to be
    
<b>superlinear</b>
    
Secant-method converges much faster than the bisection method and slower than Newton's method
    
Netwon's method , the number of floating-ponit operations takes up 2X twice as much time as the secant method in the computation of both its
    
1. function and its 
2. derivative on every iteration
    

    
Since the secant method requires only computation of its function at every step, it
    
<b>Can be considered</b> faster in absolute time
    


<big> Required that initial guess values of the secant method be close to the root,
    
Otherwise it has no guarantee of

Convergence to the solution
    
Python code for the secant method is given as follows:
   
    


In [6]:
""" Secant root-finding method """

def secant(f, a, b, tol=0.001, maxiter=100):
    """ 
    :param f: function to solve
    :param a: Initial x-axis guess value,
    :param b: Initial x-axis guess value,  where b>a
    :param tol: precision of the soltuion
    :param maxiter: Maximum number of iterations
    :return: x-axis value of the root, number of iterations used
    """ 
    
    
    
    n = 1
    while n <= maxiter:
        c = b - f(b) *((b-a) / (f(b) - f(a)))
        if abs(c-b) < tol:
                return c, n
            
        a = b
        b = c
        n+= 1
    return None, n

<big><b><i>Again,</b></i>

<b>we will <i>reuse</i> the same nonlinear function and

<b>return the results from the secant method:</b>

In [7]:
y = lambda x : x ** 3 + 2*x**2 - 5
root, iterations = secant(y, -5.0, 5.0,0.0001, 100)
print("Root is:", root)

print("Iterations:", iterations)

Root is: 1.2418965622558549
Iterations: 14


<big>Though

All of the preceding root-finding methods gave very close solutions, the

secant method

performs with fewer iterations compared to the bisection method, but with 

more than Newton's method?

## Combining Root-Finding methods
<big>Perfectly possible to 
write your own root -finding algorithms 
    
Using a <i>combination</i> of the previously mentioned root-
    
finding methods.
    
For example, <b>you may use</b> the Following implmentation in the Following order: 
    
1. USE (your brain, mind, wits, witscoin)  FASTER secant method to converge the problem to a prespecified tolerance value or a maximum number of iterations
2. Once a perspecified tolerance level is reached, switch to using the bisection method to converge to the root by halving the search interval w/ each iteration until the root is found
    
    
<b>Brent's method</b> or the <b>Wijngaarden-Dekker-Brent</b> method combines the bisection root-finding method, secant method, and 
    
<i>inverse quadratic interpolation whenever possible</i> 
    
<b>Algorithm attempts</b> to use either the secant method or inverse quadratic interpolation whenever possible, and uses the bisection method where necsesary
    
Brents' method can also be found in the <pre>scipy.optimize.brentq</pre> function of SciPy
</big>


## SciPy implementations
<big>
Before starting on writing, your root-finding algorithm to solve nonlinear or even linear probke,s

Take a look at the documentation of the <pre>scipy.optimize</pre> methods SciPy 



contains a collection of scientific computing functions as an extension of Python

Chances are these open source algorithms might fit into your applications off-the-shelf
    
</big>

## Rootfinding scalar functions
<big>Some root-
    

finding functions that can be found in the <pre>scipy.optimize</pre> modules are
    
<pre>bisect</pre>
    
<pre>newton</pre>
    
<pre>brentq</pre>
    

    
<pre>ridder</pre>
    
Let's
    
set up the examples that we have discussed using the implementations by SciPy:

In [7]:
"""
Documentation at http://docs.scipy.org/doc/scipy/reference/optimize.html
"""

import scipy.optimize as optimize

In [14]:
y = lambda x: x**3 + 2*x**2 - 5
dy= lambda x: 3.*x**2 + 4.*x


# Call method: bisection

print("Bisection method: bisect(f, a, b[, args, xtol, rtol, maxiter,...])")

print("Bisection method: %s"\
      % optimize.bisect(y, -5., 5., xtol=0.0001))
# Call method: newton(func, x0[,fprime, args, tol, ...])
print("Newton's method: %s"\
      % optimize.newton(y,  5., fprime=dy))


# When fprime=None, then the secant method is used
print("Secant method: %s" \
      % optimize.newton(y,5.))


# Call method: brentq(f, a, b[, args, xtol, rtol, maxiter,...])
print("Brent's method: %s"\
     %optimize.brentq(y, -5., 5.))

Bisection method: bisect(f, a, b[, args, xtol, rtol, maxiter,...])
Bisection method: 1.2418365478515625
Newton's method: 1.2418965630344798
Secant method: 1.2418965630344803
Brent's method: 1.241896563034559


<big>When we run the preceding code, the following output is generated:

<pre>Bisection method: bisect(f, a, b[, args, xtol, rtol, maxiter,...])
Bisection method: 1.2418365478515625
Newton's method: 1.2418965630344798
Secant method: 1.2418965630344803
Brent's method: 1.241896563034559

</pre>

<big><b>We can see</b> that the SciPy implementation gives us somewhat a very close answers as our derived ones

<big>Should be noted that SciPy ahs a set of well-defined conditions for every implementation
    
Example, the function call of the bisection routine is given as follows:
    


In [ ]:
import scipy
scipy.optimize.bisect(f,a,b, args=(), xtol=1e-12,
                     rtol=4.408920985006262e-16,
                     maxiter=100,
                     full_output=False)

<big>Function will strictly evaluate the function
<pre>f</pre> to return a zero of the function
$f(a)$ and

$f(b)$ cannot have the

same signs?

Certain scenarios: it is difficult to fulfill these constraints
    
Example, for in



Solving for nonlinear implied volatility models, 

<b>volatility values</b> cannot be negative

Active markets, 
    
finding a root or a 0 zero of the volatility function is 
    
almost impossible without modifying the underlying implementation

Co-In-side such cases implementing
    
our own root-finding
    
methods might perhaps 
    
Give us more 
    
Authority over how our
    
application should behave - AB

## General Nonlinear Solvers

<big><pre>scipy.optimize</pre> module also contains multidimensional general solvers that

<b>We CAN harness</b> to our
    
Advantage 
    
<pre>root</pre> and <pre>fsolve</pre> functions are some 
    
Examples with following function properties:
    
    
+ <pre> root(fun, x0[, args, method, ac,tol,...]):</pre>
Finds the a root of a vector function
    
+ <pre> fsolve(func, x0[, args, fprime,...]):</pre>
Finds the a root of a function
Outputs are rturend as a dictionary
    
<i>object</i>
    
<b>Using</b> our example again as inputs to these functions
    
<b>We will</b> get the following \[objective\] output
    

    

In [1]:
import scipy.optimize as optimize

y = lambda x:x**3 + 2.*x**2 -5.

dy = lambda x: 3.*x**2 + 4.*x

print(optimize.fsolve(y, 5., fprime=dy))

[1.24189656]


In [4]:
print(optimize.root(y, 5.))

    fjac: array([[-1.]])
     fun: array([3.55271368e-15])
 message: 'The solution converged.'
    nfev: 12
     qtf: array([-3.73605502e-09])
       r: array([-9.59451815])
  status: 1
 success: True
       x: array([1.24189656])


<big><b>Using</b> an initial guess value of 5,

our solution converge to the root at 1.24189656, which is pretty close to the answers we had so far

What happens when 

<b>We choose a value</b>

on the other side of the graph?

Let's use an initial guess value of -5:

In [5]:
print(optimize.fsolve(y, -5., fprime=dy))

[-1.33306553]


/home/kevin/gh/programmingbitcoin/icon-tutorial/venv/lib/python3.7/site-packages/scipy/optimize/minpack.py:175: RuntimeWarning: The iteration is not making good progress, as measured by the 
  improvement from the last ten iterations.
  warnings.warn(msg, RuntimeWarning)


In [8]:
print(optimize.root(y, -5.))

    fjac: array([[-1.]])
     fun: array([-3.81481496])
 message: 'The iteration is not making good progress, as measured by the \n  improvement from the last ten iterations.'
    nfev: 28
     qtf: array([3.81481521])
       r: array([-0.00461503])
  status: 5
 success: False
       x: array([-1.33306551])


<big>As <b>seen</b> from the display output, the algorithms 

<i>did not converge</i> and 

<b>return a root</b> that is a little further away from our previous answers

If <b>We take a look</b> at the equation on a graph, there are a numberb of points along the curve that 

<i>lie</i> very close? to the root

A root-finder would be need to <b>obtain</b> the desired



level of accuracy, while

solvers <b>attempt to solve</b> for the 

1. <i>nearest</i> answer
2. <i>fastest</i> time

## Summary

<big>Chapter, we briefly discussed the persistence of nonlinearity in economics and finance
    
We looked at some nonlinear models that are commonly used in a finance to explain certain aspects of data left unexplained by linear models: 
    
+ BS model

+ Implied volatilty model
+ Markov  switching model
+ Threshold $\tau$ model
+ Smooth transition models
    
BS implied volatility modeling, we discussed the volatility smile that was made up of implied volatilities derived via the BS model from the market prices of call or put options for a particular maturity
    
    
You may be interested enough to seek the lowest implied volatility value possible, which 
    
<b>Can be the useful</b> for co-inferring theoretical prices and comparing against the market prices for potential opportunities
    
However, since the curve is nonlinear, linear algebra cannot adequately solve for the opimal point
    
<h4>We will require the use of root-finding methods</h4>
    
Root-finding methods attempt to find the root of a function or its zero 0
We discussed common root finding methods:    
+ BS BISECTION METHOD
+ Newton's method
+ Secant method
    
Using a combination of root-finding algorithms 
<b>may help us</b> to <i>seek</i> roots of complex 
<b>functions faster</b>
    
We explored functionalities in the <pre>scipy.optimize</pre> module that contains these 
root-finding methods, albeit with constraints
One of these constraints requires that the two 2 boundary input values be 
    
Evaluated with a pair of negative value and positive value for the 
    
<b>Solution to converge successfully</b>
    

In implied volatility modeling, this evaluation is almost impossible since volatilities do not have negative values

    
Implementing our own root-finding <b>methods might perhaps </b>
    
<b>Give us more authority over how our application</b> should perform
    
<b>Using Generally Solvers</b>  is another \[new\]? way of finding roots - they may also converge to our solution moer quickly, 
    
Such a convergence is not guaranteed by the initial 
    
<i>given values</i> 
    
Nonlinear modeling and optimization are inherently a complex task, and there is 
</big>    
# $\nexists$
### No Universal Solution or a sure way to reach a conclusion
<big>This chapter serves to introduce nonlinearity studies for finance in general
    
Next chapter,
    
<b>We will</b> take a look at numerical methods commonly 
    
used for options pricing - By
    
Pairing a numerical procedure with a root-finding algorithm
    
<b>We will learn</b> <b>how to</b> an implied with the market prices of an 
    
Equity option